File: scripts/filter.js
#!/usr/bin/env node
const fs = require("fs");
const path = require("path");
const { loadConfig, getProtectedIndices, isProtected } = require("./config_manager");

const MASK = "[PROTECTED]";

function detectDelimiter(line) {
  for (const d of ["\t", ",", "|", ";"]) {
    if (line.includes(d)) return d;
  }
  return ",";
}

function parseCsvLine(line, delim) {
  const result = [];
  let current = "";
  let inQuotes = false;
  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (ch === '"') {
      if (inQuotes && i + 1 < line.length && line[i + 1] === '"') {
        current += '"';
        i++;
      } else {
        inQuotes = !inQuotes;
      }
    } else if (ch === delim && !inQuotes) {
      result.push(current);
      current = "";
    } else {
      current += ch;
    }
  }
  result.push(current);
  return result;
}

function filterCsv(text, config, delimiter) {
  const lines = text.trim().split(/\r?\n/);
  if (!lines.length) return text;
  if (!delimiter) delimiter = detectDelimiter(lines[0]);

  const headers = parseCsvLine(lines[0], delimiter);
  const protectedIdx = getProtectedIndices(headers, config);
  if (!protectedIdx.length) return text;

  const result = [lines[0]];
  for (let i = 1; i < lines.length; i++) {
    if (!lines[i].trim()) continue;
    const fields = parseCsvLine(lines[i], delimiter);
    for (const idx of protectedIdx) {
      if (idx < fields.length) fields[idx] = MASK;
    }
    result.push(fields.map((f) => (f.includes(delimiter) || f.includes('"') ? `"${f.replace(/"/g, '""')}"` : f)).join(delimiter));
  }
  return result.join("\n") + "\n";
}

function filterJson(text, config) {
  try {
    const data = JSON.parse(text);
    if (Array.isArray(data) && data.length && typeof data[0] === "object") {
      for (const record of data) {
        for (const key of Object.keys(record)) {
          if (isProtected(key, config)) record[key] = MASK;
        }
      }
      return JSON.stringify(data, null, 2);
    }
    if (typeof data === "object" && data) {
      for (const key of Object.keys(data)) {
        if (isProtected(key, config)) data[key] = MASK;
      }
      return JSON.stringify(data, null, 2);
    }
  } catch {}
  return text;
}

function filterSqlOutput(text, config) {
  const lines = text.split("\n");
  if (lines.length < 2) return text;

  let headerLine = null;
  let separatorLine = null;

  for (let i = 0; i < lines.length; i++) {
    if (/^[\s|]*-+[\s|+-]*-+/.test(lines[i])) {
      separatorLine = i;
      if (i > 0) headerLine = i - 1;
      break;
    }
  }
  if (headerLine === null) return text;

  const headerText = lines[headerLine];
  let headers, delim;
  if (headerText.includes("|")) {
    headers = headerText.split("|").map((h) => h.trim()).filter(Boolean);
    delim = "|";
  } else {
    headers = headerText.trim().split(/\s+/);
    delim = null;
  }

  const protectedIdx = getProtectedIndices(headers, config);
  if (!protectedIdx.length) return text;

  const dataStart = separatorLine !== null ? separatorLine + 1 : headerLine + 1;
  const result = [];

  for (let i = 0; i < lines.length; i++) {
    if (i <= headerLine || (separatorLine !== null && i === separatorLine)) {
      result.push(lines[i]);
    } else if (i >= dataStart && lines[i].trim() && !/^\s*\+/.test(lines[i]) && !/^\s*\(?\d+ rows?\)?/.test(lines[i])) {
      if (delim === "|") {
        const parts = lines[i].split("|");
        const realParts = [];
        let prefix = "", suffix = "";
        for (let j = 0; j < parts.length; j++) {
          if (j === 0 && !parts[j].trim()) { prefix = parts[j]; continue; }
          if (j === parts.length - 1 && !parts[j].trim()) { suffix = parts[j]; continue; }
          realParts.push(parts[j]);
        }
        for (const idx of protectedIdx) {
          if (idx < realParts.length) {
            const w = realParts[idx].length;
            realParts[idx] = w > MASK.length ? MASK.padStart((w + MASK.length) / 2).padEnd(w) : MASK;
          }
        }
        result.push(prefix + realParts.join("|") + suffix);
      } else {
        result.push(lines[i]);
      }
    } else {
      result.push(lines[i]);
    }
  }
  return result.join("\n");
}

function filterXlsx(filePath, config) {
  const zlib = require("zlib");
  const buf = fs.readFileSync(filePath);
  const entries = parseZip(buf);

  const wbEntry = entries["xl/workbook.xml"];
  if (!wbEntry) { console.log("[DATA PROTECTOR] Cannot parse xlsx"); return; }
  const wbXml = zlib.inflateRawSync(wbEntry).toString("utf-8");
  const sheetNames = [];
  const sheetRe = /<sheet\s[^>]*name="([^"]*)"[^>]*\/>/g;
  let m;
  while ((m = sheetRe.exec(wbXml))) sheetNames.push(m[1]);

  const ssEntry = entries["xl/sharedStrings.xml"];
  const strings = [];
  if (ssEntry) {
    const ssXml = zlib.inflateRawSync(ssEntry).toString("utf-8");
    const siRe = /<si>([\s\S]*?)<\/si>/g;
    let sm;
    while ((sm = siRe.exec(ssXml))) {
      const tRe = /<t[^>]*>([\s\S]*?)<\/t>/g;
      let parts = [];
      let tm;
      while ((tm = tRe.exec(sm[1]))) parts.push(tm[1]);
      strings.push(parts.join(""));
    }
  }

  for (let i = 0; i < sheetNames.length; i++) {
    const sheetFile = entries[`xl/worksheets/sheet${i + 1}.xml`];
    if (!sheetFile) continue;
    const sheetXml = zlib.inflateRawSync(sheetFile).toString("utf-8");

    const rows = [];
    const rowRe = /<row\b[^>]*>([\s\S]*?)<\/row>/g;
    let rm;
    while ((rm = rowRe.exec(sheetXml))) {
      const cells = [];
      const cellRe = /<c\s([^>]*)>([\s\S]*?)<\/c>|<c\s([^>]*)\/>/g;
      let cm;
      while ((cm = cellRe.exec(rm[1]))) {
        const attrs = cm[1] || cm[3] || "";
        const inner = cm[2] || "";
        const typeMatch = attrs.match(/t="([^"]*)"/);
        const valMatch = inner.match(/<v>([\s\S]*?)<\/v>/);
        const inlineMatch = inner.match(/<is>[\s\S]*?<t[^>]*>([\s\S]*?)<\/t>[\s\S]*?<\/is>/);
        if (typeMatch && typeMatch[1] === "inlineStr" && inlineMatch) {
          cells.push(inlineMatch[1]);
        } else if (valMatch) {
          if (typeMatch && typeMatch[1] === "s") {
            cells.push(strings[parseInt(valMatch[1])] || "");
          } else {
            cells.push(valMatch[1]);
          }
        } else {
          cells.push("");
        }
      }
      rows.push(cells);
    }

    if (!rows.length) continue;
    const headers = rows[0];
    const protectedIdx = getProtectedIndices(headers, config);

    console.log(`=== Sheet: ${sheetNames[i]} (${rows.length - 1} rows x ${headers.length} cols) ===`);
    console.log(headers.join("\t"));
    for (let r = 1; r < rows.length; r++) {
      const vals = rows[r].map((v, ci) => (protectedIdx.includes(ci) ? MASK : v));
      console.log(vals.join("\t"));
    }
    console.log();
  }
}

function parseZip(buf) {
  const entries = {};
  let pos = buf.length - 22;
  while (pos >= 0 && buf.readUInt32LE(pos) !== 0x06054b50) pos--;
  if (pos < 0) return entries;
  const cdOffset = buf.readUInt32LE(pos + 16);
  let p = cdOffset;
  while (p < pos && buf.readUInt32LE(p) === 0x02014b50) {
    const compSize = buf.readUInt32LE(p + 20);
    const nameLen = buf.readUInt16LE(p + 28);
    const extraLen = buf.readUInt16LE(p + 30);
    const commentLen = buf.readUInt16LE(p + 32);
    const localOffset = buf.readUInt32LE(p + 42);
    const name = buf.toString("utf-8", p + 46, p + 46 + nameLen);
    const lp = localOffset;
    const lNameLen = buf.readUInt16LE(lp + 26);
    const lExtraLen = buf.readUInt16LE(lp + 28);
    const dataStart = lp + 30 + lNameLen + lExtraLen;
    if (compSize > 0) entries[name] = buf.slice(dataStart, dataStart + compSize);
    p += 46 + nameLen + extraLen + commentLen;
  }
  return entries;
}

function filterFile(filePath, config) {
  const ext = path.extname(filePath).toLowerCase();
  if (ext === ".xlsx" || ext === ".xls") {
    filterXlsx(filePath, config);
    return;
  }
  if (ext === ".parquet") {
    console.error("[DATA PROTECTOR] Parquet not supported directly.");
    process.exit(1);
  }
  const content = fs.readFileSync(filePath, "utf-8");
  if (ext === ".json") console.log(filterJson(content, config));
  else if (ext === ".csv") console.log(filterCsv(content, config, ","));
  else if (ext === ".tsv") console.log(filterCsv(content, config, "\t"));
  else console.log(filterCsv(content, config));
}

function filterStdin(config, fmt) {
  let input = "";
  process.stdin.setEncoding("utf-8");
  process.stdin.on("data", (chunk) => (input += chunk));
  process.stdin.on("end", () => {
    if (!input.trim()) { process.stdout.write(input); return; }
    if (fmt === "sql") console.log(filterSqlOutput(input, config));
    else if (fmt === "json") console.log(filterJson(input, config));
    else {
      try { JSON.parse(input); console.log(filterJson(input, config)); return; } catch {}
      if (/^[\s|]*-+[\s|+-]*-+/m.test(input)) console.log(filterSqlOutput(input, config));
      else console.log(filterCsv(input, config));
    }
  });
}

function main() {
  let filePath = null;
  const args = process.argv.slice(2);

  for (let i = 0; i < args.length; i++) {
    if (args[i] !== "--stdin" && args[i] !== "--format") {
      filePath = args[i];
      break;
    }
    if (args[i] === "--format") i++;
  }

  const config = loadConfig(filePath || undefined);

  if (args.includes("--stdin")) {
    let fmt = null;
    const fmtIdx = args.indexOf("--format");
    if (fmtIdx >= 0 && fmtIdx + 1 < args.length) fmt = args[fmtIdx + 1];
    filterStdin(config, fmt);
  } else if (filePath) {
    filterFile(filePath, config);
  } else {
    console.error("Usage: filter.js <file_path> | filter.js --stdin [--format sql|json]");
    process.exit(1);
  }
}

main();

File: scripts/filter.py
#!/usr/bin/env python3
"""Core column filtering engine for data-protector plugin.

Usage:
  filter.py <file_path>          Read file, output with protected columns masked
  filter.py --stdin              Read stdin (piped output), filter columns
  filter.py --stdin --format sql Parse SQL-formatted output
"""
import csv
import io
import json
import os
import re
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from config_manager import load_config, get_protected_indices, is_protected

MASK = "[PROTECTED]"

DATA_EXTENSIONS = {".csv", ".tsv", ".xlsx", ".xls", ".json", ".parquet"}


def detect_delimiter(line):
    for delim in ["\t", ",", "|", ";"]:
        if delim in line:
            return delim
    return ","


def filter_csv(text, config, delimiter=None):
    lines = text.strip().split("\n")
    if not lines:
        return text

    if delimiter is None:
        delimiter = detect_delimiter(lines[0])

    reader = csv.reader(io.StringIO(text), delimiter=delimiter)
    rows = list(reader)
    if not rows:
        return text

    headers = rows[0]
    protected = get_protected_indices(headers, config)
    if not protected:
        return text

    out = io.StringIO()
    writer = csv.writer(out, delimiter=delimiter)
    for i, row in enumerate(rows):
        if i == 0:
            writer.writerow(row)
        else:
            masked = list(row)
            for idx in protected:
                if idx < len(masked):
                    masked[idx] = MASK
            writer.writerow(masked)
    return out.getvalue()


def filter_json_data(text, config):
    try:
        data = json.loads(text)
    except json.JSONDecodeError:
        return text

    if isinstance(data, list) and data and isinstance(data[0], dict):
        for record in data:
            for key in list(record.keys()):
                if is_protected(key, config):
                    record[key] = MASK
        return json.dumps(data, indent=2, ensure_ascii=False)
    elif isinstance(data, dict):
        for key in list(data.keys()):
            if is_protected(key, config):
                data[key] = MASK
        return json.dumps(data, indent=2, ensure_ascii=False)
    return text


def filter_sql_output(text, config):
    lines = text.split("\n")
    if len(lines) < 2:
        return text

    header_line = None
    separator_line = None
    header_idx = -1

    for i, line in enumerate(lines):
        if re.match(r"^[\s|]*-+[\s|+-]*-+", line):
            separator_line = i
            if i > 0:
                header_line = i - 1
            break
        if "|" in line and not re.match(r"^\s*\+", line):
            potential_headers = [h.strip() for h in line.split("|") if h.strip()]
            if len(potential_headers) >= 2 and all(re.match(r"^[a-zA-Z_]\w*$", h) for h in potential_headers):
                header_line = i
                header_idx = i

    if header_line is None:
        return text

    header_text = lines[header_line]
    if "|" in header_text:
        headers = [h.strip() for h in header_text.split("|")]
        headers = [h for h in headers if h]
        delim = "|"
    else:
        headers = header_text.split()
        delim = None

    protected = get_protected_indices(headers, config)
    if not protected:
        return text

    result = []
    data_start = (separator_line + 1) if separator_line is not None else (header_line + 1)

    for i, line in enumerate(lines):
        if i <= header_line or (separator_line is not None and i == separator_line):
            result.append(line)
        elif i >= data_start and line.strip() and not re.match(r"^\s*\+", line) and not re.match(r"^\s*\(?\d+ rows?\)?", line):
            if delim == "|":
                parts = line.split("|")
                real_parts = []
                prefix = ""
                suffix = ""
                for j, p in enumerate(parts):
                    if j == 0 and not p.strip():
                        prefix = p
                        continue
                    if j == len(parts) - 1 and not p.strip():
                        suffix = p
                        continue
                    real_parts.append(p)

                for idx in protected:
                    if idx < len(real_parts):
                        orig = real_parts[idx]
                        width = len(orig)
                        real_parts[idx] = MASK.center(width) if width > len(MASK) else MASK

                result.append(prefix + "|".join(real_parts) + suffix)
            else:
                result.append(line)
        else:
            result.append(line)

    return "\n".join(result)


def filter_cat_numbered(text, config):
    """Handle `cat -n` style output (line numbers + tab + content)."""
    lines = text.split("\n")
    if not lines:
        return text

    numbered_pattern = re.compile(r"^(\s*\d+\t)(.*)$")
    first_content = None
    for line in lines:
        m = numbered_pattern.match(line)
        if m:
            first_content = m.group(2)
            break
        elif line.strip():
            first_content = line
            break

    if first_content is None:
        return text

    delimiter = detect_delimiter(first_content)
    is_numbered = bool(numbered_pattern.match(lines[0])) if lines[0].strip() else False

    if not is_numbered:
        return filter_csv(text, config, delimiter)

    content_lines = []
    prefixes = []
    for line in lines:
        m = numbered_pattern.match(line)
        if m:
            prefixes.append(m.group(1))
            content_lines.append(m.group(2))
        else:
            prefixes.append("")
            content_lines.append(line)

    filtered_content = filter_csv("\n".join(content_lines), config, delimiter)
    filtered_lines = filtered_content.split("\n")

    result = []
    for i, fline in enumerate(filtered_lines):
        prefix = prefixes[i] if i < len(prefixes) else ""
        result.append(prefix + fline)
    return "\n".join(result)


def filter_xlsx(file_path, config):
    try:
        import openpyxl
    except ImportError:
        import subprocess
        result = subprocess.run(
            [sys.executable, "-m", "pip", "install", "--quiet", "openpyxl"],
            capture_output=True
        )
        import openpyxl

    wb = openpyxl.load_workbook(file_path, read_only=True, data_only=True)
    for sheet_name in wb.sheetnames:
        ws = wb[sheet_name]
        rows = list(ws.iter_rows(values_only=True))
        if not rows:
            continue

        headers = [str(h) if h is not None else "" for h in rows[0]]
        protected = get_protected_indices(headers, config)

        print(f"=== Sheet: {sheet_name} ({len(rows)-1} rows x {len(headers)} cols) ===")
        print("\t".join(headers))
        for row in rows[1:]:
            vals = []
            for i, v in enumerate(row):
                if i in protected:
                    vals.append(MASK)
                else:
                    vals.append(str(v) if v is not None else "")
            print("\t".join(vals))
        print()
    wb.close()


def filter_file(file_path, config):
    ext = os.path.splitext(file_path)[1].lower()

    if ext in (".xlsx", ".xls"):
        filter_xlsx(file_path, config)
        return

    if ext == ".parquet":
        print(f"[DATA PROTECTOR] Parquet not supported directly. Use pandas to read.", file=sys.stderr)
        sys.exit(1)

    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()

    if ext == ".json":
        print(filter_json_data(content, config))
    elif ext in (".csv", ".tsv"):
        delimiter = "\t" if ext == ".tsv" else ","
        print(filter_csv(content, config, delimiter))
    else:
        print(filter_csv(content, config))


def filter_stdin(config, fmt=None):
    text = sys.stdin.read()
    if not text.strip():
        print(text, end="")
        return

    if fmt == "sql":
        print(filter_sql_output(text, config))
    elif fmt == "json":
        print(filter_json_data(text, config))
    else:
        try:
            json.loads(text)
            print(filter_json_data(text, config))
            return
        except (json.JSONDecodeError, ValueError):
            pass

        if re.search(r"^[\s|]*-+[\s|+-]*-+", text, re.MULTILINE):
            print(filter_sql_output(text, config))
        else:
            print(filter_cat_numbered(text, config))


def main():
    file_path = None
    if len(sys.argv) > 1 and sys.argv[1] != "--stdin":
        file_path = sys.argv[1]

    config = load_config(file_path)

    if "--stdin" in sys.argv:
        fmt = None
        if "--format" in sys.argv:
            fmt_idx = sys.argv.index("--format") + 1
            if fmt_idx < len(sys.argv):
                fmt = sys.argv[fmt_idx]
        filter_stdin(config, fmt)
    elif file_path:
        filter_file(file_path, config)
    else:
        print("Usage: filter.py <file_path> | filter.py --stdin [--format sql|json]", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
